lib/spack/env/cc

```diff
@@ -114,104 +114,118 @@ case "$command" in
 esac
 
 # Finish setting up the mode.
-if [ -z "$mode" ]; then
+if [[ -z $mode ]]; then
     mode=ccld
     for arg in "$@"; do
-        if [ "$arg" = -v -o "$arg" = -V -o "$arg" = --version -o "$arg" = -dumpversion ]; then
-            mode=vcheck
-            break
-        elif [ "$arg" = -E ]; then
-            mode=cpp
-            break
-        elif [ "$arg" = -c ]; then
-            mode=cc
-            break
-        fi
+        case "$arg" in
+            -v|-V|--version|-dumpversion)
+                mode=vcheck
+                break
+                ;;
+            -E)
+                mode=cpp
+                break
+                ;;
+            -c)
+                mode=cc
+                break
+                ;;
+            -S)
+                mode=as
+                break
+                ;;
+        esac
     done
 fi
 
 # Dump the version and exist if we're in testing mode.
-if [ "$SPACK_TEST_COMMAND" = "dump-mode" ]; then
+if [[ $SPACK_TEST_COMMAND = dump-mode ]]; then
     echo "$mode"
     exit
 fi
 
 # Check that at least one of the real commands was actually selected,
 # otherwise we don't know what to execute.
-if [ -z "$command" ]; then
+if [[ -z $command ]]; then
     die "ERROR: Compiler '$SPACK_COMPILER_SPEC' does not support compiling $language programs."
 fi
 
 # Save original command for debug logging
 input_command="$@"
+args=("$@")
 
-#
-# Now do real parsing of the command line args, trying hard to keep
-# non-rpath linker arguments in the proper order w.r.t. other command
-# line arguments.  This is important for things like groups.
-#
-includes=()
-libraries=()
-libs=()
-rpaths=()
-other_args=()
+# Dump parsed values for unit testing if asked for
+if [[ -n $SPACK_TEST_COMMAND ]]; then
```

Comment:
Oh, because you are no longer using those arrays outside of that if-statement. Hmm, I'm not sure I understand exactly how this code works (before or after your changes). It feels like we're collecting the same parameters twice.